pkg/controller/networkpolicy/tier.go

```diff
@@ -0,0 +1,165 @@
+		{
+			ObjectMeta: metav1.ObjectMeta{
+				Name: "application",
+			},
+			Spec: secv1alpha1.TierSpec{
+				Priority:    priorityMap["application"],
+				Description: "[READ-ONLY]: System generated default Application Tier",
+			},
+		},
+		{
+			ObjectMeta: metav1.ObjectMeta{
+				Name: "platform",
+			},
+			Spec: secv1alpha1.TierSpec{
+				Priority:    priorityMap["platform"],
+				Description: "[READ-ONLY]: System generated Platform Tier",
+			},
+		},
+		{
+			ObjectMeta: metav1.ObjectMeta{
+				Name: "networkops",
+			},
+			Spec: secv1alpha1.TierSpec{
+				Priority:    priorityMap["networkops"],
+				Description: "[READ-ONLY]: System generated NetworkOps Tier",
+			},
+		},
+		{
+			ObjectMeta: metav1.ObjectMeta{
+				Name: "securityops",
+			},
+			Spec: secv1alpha1.TierSpec{
+				Priority:    priorityMap["securityops"],
+				Description: "[READ-ONLY]: System generated SecurityOps Tier",
+			},
+		},
+		{
+			ObjectMeta: metav1.ObjectMeta{
+				Name: "emergency",
+			},
+			Spec: secv1alpha1.TierSpec{
+				Priority:    priorityMap["emergency"],
+				Description: "[READ-ONLY]: System generated Emergency Tier",
+			},
+		},
+	}
+)
+
+// InitializeTiers initializes the default Tiers created by Antrea on init. It
+// will first attempt to retrieve the Tier by it's name from K8s and if missing,
+// create the CR. InitializeTiers will be called as part of a Post-Start hook
+// of antrea-controller's APIServer.
+func (n *NetworkPolicyController) InitializeTiers() {
+	for _, t := range systemGeneratedTiers {
+		// Check if Tier is already present.
+		tr, err := n.tierLister.Get(t.Name)
+		if tr != nil && err == nil {
+			// Tier is already present.
+			klog.Infof("%s Tier already created", t.Name)
+			continue
+		}
+		err = n.initTier(t)
+		if err != nil {
+			// System generated Tiers were not initialized within an acceptable delay.
+			klog.Errorf("Failed to create system Tier %s on init", t.Name)
+		}
+	}
+}
+
+func (n *NetworkPolicyController) initTier(t *secv1alpha1.Tier) error {
+	var err error
+	for i := 1; i <= retryInitTier; i++ {
+		// Allow APIServer to start and accept requests for Tier CREATE validation.
+		time.Sleep(retryInitInterval)
+		klog.Infof("Creating %s Tier", t.Name)
+		_, err = n.crdClient.SecurityV1alpha1().Tiers().Create(context.TODO(), t, metav1.CreateOptions{})
+		if err != nil {
+			klog.Warningf("Failed to create %s Tier on init: %v. Retry attempt: %d", t.Name, err, i)
+			// Tier creation may fail because antrea APIService is not yet ready
+			// to accept requests for validation. Retry fixed number of times
+			// not exceeding 2 * 5 = 10s.
+			continue
+		} else {
+			return nil
+		}
+	}
+	return err
+}
+
+// addTier receives Tier ADD events and updates the TierPrioritySet.
+func (n *NetworkPolicyController) addTier(obj interface{}) {
+	t := obj.(*secv1alpha1.Tier)
+	klog.Infof("Processing Tier %s ADD event", t.Name)
+	// Insert Tier's Priority in the unique set.
+	n.tierPrioritySet.Insert(t.Spec.Priority)
```

Comment:
I wonder if tierPrioritySet is reliable enough to check the uniqueness as informer is asynchronous.
For example, if creating two tiers with same priority at the same time, both requests might succeed if the second is handled before the controller receives and handles the add event of the first object.
Have you tried it? But I don't know any better way either if the issue is real.

Reply:
good point.. it is real issue.. do you think we should consider my previous approach of using the indexer here instead of sets?

Reply:
using indexer did not allow overlapping priorities..

Reply:
I guess using indexer reduces the possibility because the indexer is updated earlier than handler processing events. But it's still asynchronous so there might be a chance that the second tier is created before controller receives the event of the first tier. If your test succeeded consistently, then maybe the possibility is small enough.

Reply:
yes.. i did some more tests.. and i had 1 occurrence of the race today .. on the other hand with sets.. it was almost every attempt was a failure.. I'm not sure too what's a fool proof way here :(